Add health check

bargom opened the pull request:
Added a healthCheck function, that will start to execute when the
circuit is open. If the function is defined, it will overwrite the
timeout function which is automatically closing the circuit after
{circuitDuration} ms. The health check is used to check the external
system each {healthCheckInterval} and close the circuit only when the
external system is responding.

Example:
[testCircuitBreaker3.js.txt](https://github.com/awolden/brakes/files/346304/testCircuitBreaker3.js.txt)

bargom commented:
If this function is a good idea, I can work on more coverage :)

awolden commented:
Hey @bargom,

This looks like a solid addition to the module! I like the concept, and it looks like it can be implemented without much refactoring or disruption of the current API.

There are a couple tweaks I would ask for before merging the PR:

- [ ] Update readme to include new functionality.

- [ ] Make it so the health function can also be passed in via the options object. I.e.:

```javascript
if (this._opts.healthCheck) this.healthCheck(this._opts.healthCheck);
```

- [ ] Refactor `_open()` to reduce logic. With this healthCheck addition I feel there is a bit too much logic in that function, and it will make achieving appropriate test coverage more difficult. The interval and timeout logic should be pulled out and placed in separate functions:

```javascript
_open() {
  if (this._circuitOpen) return;
  this.emit('circuitOpen');
  this._circuitOpen = true;
  if (this._healthCheck){
   this._setHealthInterval();
  }
  else {
   this._resetCircuitTimeout();
  }
}

_setHealthInterval(){
  const interval = setInterval(() => {
    this._healthCheck().then(() => {
      this._stats.reset();
      this._close();
      clearInterval(interval);
    })
  }, this._opts.healthCheckInterval);
  interval.unref();
}

_resetCircuitTimeout(){
  const timer = setTimeout(() => {
    this._stats.reset();
    this._close();
  }, this._opts.circuitDuration);
  timer.unref();
}
```

Thanks for your interest in contributing to the library! If you have any questions or need help with anything else concerning Brakes, let me know here or in the issues page. 

-Alex Wolden

bargom commented:
Hello @awolden,

Now I finished the changes, updated files, readme, and tests, added more examples and committed/pushed.

There are some more code in the `_setHealthInterval` now, I noticed some flows in the logic while running the `healthCheck-example.js`.

```javascript
1 _setHealthInterval() {
2     const interval = setInterval(() => {
3       if (this._circuitOpen) {
4         this._healthCheck().then(() => {
5           // it is possible that in the meantime, the circuit is already
6           // closed by the previous health check
7            if (this._circuitOpen) {
8              this._stats.reset();
9              this._close();
10           }
11          clearInterval(interval);
12         }).catch(err => {
13           this.emit('healthCheckFailed', err);
14         });
15       }
16       else {
            // the circuit is closed out of health check,
            // or from one of the cascading health checks
            // (if the interval is not long enough to wait for one
            // health check to complete, the previous health check might
            // close the circuit) OR (manually closed).
17         clearInterval(interval);
      }
    }, this._opts.healthCheckInterval);
    interval.unref();
  }
```

**Line 3:** Added the check for circuit open, because it is possible that while calling the function, the circuit might be closed manually.
**Line 7:** Extra check because if the health check is taking longer than the interval, it is possible that the previous call might close the circuit already.
**Line 17:** If the circuit is already closed, here is the termination of interval call by calling the clearInterval.

The only difficulty I am having, is to test the health check, any ideas?

awolden commented:
Everything looks solid @bargom! Thanks for the great work on this. Testing the internals of a `setInterval` or `setTimeout` can be tricky. The general strategy here would be to stub out all of the functions it calls using sinon, then call the function, wait a determined amount of time, and verify that the stubbed functions were called.

If you want I can finish up the testing of this functionality, but this PR would likely have to be closed and a new one would be opened pointing to the branch: `feature/health-check`

jquatier commented:
I'd recommend using the fake timers in Sinon for testing the `setTimeout` and `setInterval` methods: http://sinonjs.org/docs/#clock

bargom commented:
Thanks for the comments @awolden. It would be nice if you can take over the testing functionality. If you can have one test that can handle the `setTimeout` or `setInterval`, I can also continue on the new branch. This PR can be closed then.
